File: java/02-two-pointers/012-trapping-rain-water.java
/*
 * LeetCode #42 - Trapping Rain Water
 *
 * Pattern:
 * Two Pointers + Greedy (Process Smaller Boundary)
 *
 * Approach:
 * 1. Use two pointers: left and right.
 * 2. Maintain leftMax and rightMax.
 * 3. Always process the side with the smaller max boundary.
 * 4. Update the boundary first, then calculate trapped water.
 *
 * Key Insight:
 * Water level at any index is determined by:
 * min(maxLeft, maxRight) - height[i]
 * We process the side with smaller max boundary because
 * that side limits the water level.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }

        int left = 0;
        int right = height.length - 1;

        int leftMax = height[left];
        int rightMax = height[right];

        int water = 0;

        while (left < right) {

            if (leftMax < rightMax) {
                left++;
                leftMax = Math.max(leftMax, height[left]);
                water += leftMax - height[left];
            } else {
                right--;
                rightMax = Math.max(rightMax, height[right]);
                water += rightMax - height[right];
            }
        }

        return water;
    }
}

File: python/02-two-pointers/012-trapping-rain-water.py
"""
LeetCode #42 - Trapping Rain Water

Pattern:
Two Pointers + Greedy (Process smaller boundary)

Approach:
1. Use two pointers: left and right.
2. Maintain leftMax and rightMax.
3. Always process the side with smaller max boundary.
4. Add trapped water accordingly.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        left = 0
        right = len(height) - 1

        leftMax = height[left]
        rightMax = height[right]

        water = 0

        while left < right:

            if leftMax < rightMax:
                left += 1
                leftMax = max(leftMax, height[left])
                water += leftMax - height[left]
            else:
                right -= 1
                rightMax = max(rightMax, height[right])
                water += rightMax - height[right]

        return water
